sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -499,6 +509,189 @@ def reader(self, test_bigquery_client=None):
+      self.use_legacy_sql = not use_standard_sql
+      self.table_reference = None
+
+    self.gcs_location = gcs_location
+    self.project = project
+    self.validate = validate
+    self.flatten_results = flatten_results
+    self.coder = coder or _JsonToDictCoder
+    self.kms_key = kms_key
+    self.split_result = None
+
+  def estimate_size(self):
+    bq = bigquery_tools.BigQueryWrapper()
+    if self.table_reference is not None:
+      table = bq.get_table(self.table_reference.projectId,
+                           self.table_reference.datasetId,
+                           self.table_reference.tableId)
+      return int(table.numBytes)
+    else:
+      self._setup_temporary_dataset(bq)
+      job = bq._start_query_job(self.project, self.query,
+                                self.use_legacy_sql, self.flatten_results,
+                                job_id=uuid.uuid4().hex, dry_run=True,
+                                kms_key=self.kms_key)
+      size = int(job.statistics.totalBytesProcessed)
+
+      bq.clean_up_temporary_dataset(self.project)
+
+      return size
+
+  def split(self, desired_bundle_size, start_position=None, stop_position=None):
+    if self.split_result is None:
+      bq = bigquery_tools.BigQueryWrapper()
+
+      if self.query is not None:
+        self._setup_temporary_dataset(bq)
+        self.table_reference = self._execute_query(bq)
+
+      schema, metadata_list = self._export_files(bq)
+      self.split_result = [TextSource(metadata.path, 0,
+                                      CompressionTypes.UNCOMPRESSED, True,
+                                      self.coder(schema))
+                           for metadata in metadata_list]
+
+      if self.query is not None:
+        bq.clean_up_temporary_dataset(self.project)
+
+    for source in self.split_result:
+      yield SourceBundle(0, source, None, None)
+
+  def get_range_tracker(self, start_position, stop_position):
+    raise NotImplementedError('BigQuery source must be split before being read')
+
+  def read(self, range_tracker):
+    raise NotImplementedError('BigQuery source must be split before being read')
+
+  def _setup_temporary_dataset(self, bq):
+    location = bq.get_query_location(self.project, self.query,
+                                     self.use_legacy_sql)
+    bq.create_temporary_dataset(self.project, location)
+
+  def _execute_query(self, bq):
+    job = bq._start_query_job(self.project, self.query,
+                              self.use_legacy_sql, self.flatten_results,
+                              job_id=uuid.uuid4().hex, kms_key=self.kms_key)
+    job_ref = job.jobReference
+    bq.wait_for_bq_job(job_ref)
+    return bq._get_temp_table(self.project)
+
+  def _export_files(self, bq):
+    """Runs a BigQuery export job.
+
+    Returns:
+      bigquery.TableSchema instance, a list of FileMetadata instances
+    """
+    job_id = uuid.uuid4().hex
+    job_ref = bq.perform_extract_job([self.gcs_location], job_id,
```

Comment:
We should consider the possibility of failures:
If this call fails and is retried - but an export job still runs twice, we may duplicate the files for export. Is that a correct assessment? 
If duplication of files on retries is a possibility, then we should generate a random subdirectory within `gcs_location` to only read from that subdirectory when the call succeeds. Does that make sense?

Reply:
I found in the doc[1] something that may answer your question:
```
Calling jobs.insert on a given job ID is idempotent; in other words, you can retry as many times as you like on the same job ID, and at most one of those operations will succeed.
```
When retry is called, job ID is the same, so I think we are save.

[1] https://cloud.google.com/bigquery/docs/exporting-data#exporting_data_stored_in

Reply:
Except the job id is not the same, right? because `uuid.uuid4().hex` returns different IDs every time?

Reply:
`_export_files()` is called only once, so the job id is the same within a pipeline. And in the case when two or more pipelines are ran at the same time, we generate a random subdirectory within `gcs_location` right here:

https://github.com/apache/beam/blob/894431000cf72a122422257365e0f3b775c0c44f/sdks/python/apache_beam/io/gcp/bigquery.py#L1571-L1572

Reply:
gotcha. This makes sense. Thanks Kamil!